src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/HaFlowMapper.java

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2023 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowmonitoring.mapper;
+
+import static java.lang.String.format;
+
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaFlowPath;
+import org.openkilda.model.HaSubFlow;
+import org.openkilda.model.PathSegment;
+import org.openkilda.wfm.topology.flowmonitoring.model.FlowState;
+import org.openkilda.wfm.topology.flowmonitoring.model.Link;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Mapper
+public class HaFlowMapper {
+    public static final HaFlowMapper INSTANCE = Mappers.getMapper(HaFlowMapper.class);
+
+    /**
+     * Convert HA-Sub-Flow path representation.
+     */
+    public FlowState toFlowStates(HaSubFlow haSubFlow) {
+        HaFlow haFlow = haSubFlow.getHaFlow();
+
+        if (haFlow != null) {
+            FlowState flowState = new FlowState();
+            flowState.setForwardPath(toLinks(getPathSegments(haFlow.getForwardPath(), haSubFlow.getHaSubFlowId())));
+            flowState.setReversePath(toLinks(getPathSegments(haFlow.getReversePath(), haSubFlow.getHaSubFlowId())));
+            return flowState;
```

Comment:
consider adding a builder to FlowState so you can just create the new object in one step instead of three

Reply:
It is a shared class and adding a builder creates an issue with backward compatibility. I would leave it as it is. 
Thanks for the idea!

Reply:
Which issue? Adding a builder doesn't change anything. if it's necessary, the constructor can continue to be used.

Reply:
If you take a look at the FlowState class, you see that we have initialized fields there. Builder changes this logic. Obviously, we can refactor such behavior and rewrite related logic, but I believe that the risk to create an additional bug in the existing implementation doesn't worth the just different code look. I mean with the builder it would look like this:
`return FlowState.builder()
            .setForwardPath(toLinks(getPathSegments(haFlow.getForwardPath(), haSubFlow.getHaSubFlowId())))
            .setReversePath(toLinks(getPathSegments(haFlow.getReversePath(), haSubFlow.getHaSubFlowId())))
            .build();`